controllers/druid/supervisorspec_controller.go

```diff
@@ -0,0 +1,608 @@
+	DetailedState string          `json:"detailedState"`
+	Healthy       bool            `json:"healthy"`
+	Spec          json.RawMessage `json:"spec"`
+	Suspended     bool            `json:"suspended"`
+}
+
+type DruidApiSupervisorSpecMinDataSchema struct {
+	DataSource string `json:"dataSource"`
+}
+
+type DruidApiSupervisorSpecMin struct {
+	DataSchema DruidApiSupervisorSpecMinDataSchema `json:"dataSchema"`
+}
+
+type SupervisorSpecStateEntry struct {
+	Id                   string `json:"id"`
+	DruidServiceEndpoint string `json:"druidHost"`
+}
+
+const (
+	ActionCreate                  = "create"
+	ActionUpdate                  = "update"
+	ActionDelete                  = "delete"
+	SupervisorSpecConfigMap       = "supervisor-specs-controller"
+	SupervisorListUrlPattern      = "http://%s:8088/druid/indexer/v1/supervisor"
+	SupervisorUrlPattern          = "http://%s:8088/druid/indexer/v1/supervisor/%s"
+	SupervisorTerminateUrlPattern = "http://%s:8088/druid/indexer/v1/supervisor/%s/terminate"
+	SupervisorResumeUrlPattern    = "http://%s:8088/druid/indexer/v1/supervisor/%s/resume"
+	SupervisorSuspendUrlPattern   = "http://%s:8088/druid/indexer/v1/supervisor/%s/suspend"
+)
+
+// SupervisorSpecReconciler reconciles a SupervisorSpec object
+type SupervisorSpecReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=druid.apache.org,resources=supervisorspecs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=druid.apache.org,resources=supervisorspecs/status,verbs=get;update;patch
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *SupervisorSpecReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&druidv1alpha1.SupervisorSpec{}).
+		Complete(r)
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the SupervisorSpec object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
+func (r *SupervisorSpecReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithName(string(uuid.NewUUID()))
+	_ = log.WithValues("supervisorspec", req.NamespacedName)
+
+	log.Info(fmt.Sprintf("reconciling SupervisorSpec (%s)", req.NamespacedName.String()))
+
+	action := ActionCreate
+	supervisorCr := &druidv1alpha1.SupervisorSpec{}
+	err := r.Client.Get(ctx, req.NamespacedName, supervisorCr)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			log.Error(err, "failed to get SupervisorSpec from k8s api")
+		}
+		action = ActionDelete
```

Comment:
If I get it right, this is how you handle Kubernetes object deletion? You shouldn't handle it that way. You should check if Kubernetes added to this object a `DeletionTimestamp`.

Also, what happens if the Druid cluster is deleted?